suppress all Autograd keys in AutoNonVariableTypeMode

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42658 Include/ExcludeDispatchKeySetGuard API
* #42619 update DispatchKey::toString()
* **#42610 suppress all Autograd keys in AutoNonVariableTypeMode**

Fix for https://github.com/pytorch/pytorch/issues/42609: `AutoNonVariableTypeMode` should suppress all autograd dispatch keys, not just `Autograd` (e.g. `XLAPreAutograd`, `PrivateUse<N>_PreAutograd`)

Differential Revision: [D22963408](https://our.internmc.facebook.com/intern/diff/D22963408)

Commit:
suppress XLAPreAutograd in AutoNonVariableTypeMode

[ghstack-poisoned]

ezyang commented:
You should also do the reserved use private autograd keys too. (Also, probably should just code the RAII guard, so you don't have to keep banging on TLS separately for each ctor/dtor)

Commit:
Update on "suppress XLAPreAutograd in AutoNonVariableTypeMode"




[ghstack-poisoned]

bhosmer commented:
@ezyang updated

Commit:
Update on "suppress all Autograd keys in AutoNonVariableTypeMode"


Fix for https://github.com/pytorch/pytorch/issues/42609: `AutoNonVariableTypeMode` should suppress all autograd dispatch keys, not just `Autograd` (e.g. `XLAPreAutograd`, `PrivateUse<N>_PreAutograd`)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/LegacyTypeDispatch.h:
```diff
@@ -43,15 +43,26 @@ namespace at {
 // trace).  To unify the two, we would first have to move profiling and tracing
 // out of VariableType.
 
+// TODO centralize this as part of dispatch key revamp
```
YUPP. Probably somewhere in like c10 headers in the end.

ezyang commented on c10/core/impl/LocalDispatchKeySet.h:
```diff
@@ -93,6 +93,21 @@ class C10_API ExcludeDispatchKeyGuard {
   bool prev_state_;
 };
 
+class C10_API ExcludeDispatchKeySetGuard {
```
Maybe we can just get rid of ExcludeDispatchKeyGuard now, only support sets.

bhosmer replied:
Went ahead and did this, for Include as well.

bhosmer replied:
#42658

Commit:
Update on "suppress all Autograd keys in AutoNonVariableTypeMode"


Fix for https://github.com/pytorch/pytorch/issues/42609: `AutoNonVariableTypeMode` should suppress all autograd dispatch keys, not just `Autograd` (e.g. `XLAPreAutograd`, `PrivateUse<N>_PreAutograd`)

Differential Revision: [D22963408](https://our.internmc.facebook.com/intern/diff/D22963408)

[ghstack-poisoned]

Commit:
Update on "suppress all Autograd keys in AutoNonVariableTypeMode"


Fix for https://github.com/pytorch/pytorch/issues/42609: `AutoNonVariableTypeMode` should suppress all autograd dispatch keys, not just `Autograd` (e.g. `XLAPreAutograd`, `PrivateUse<N>_PreAutograd`)

Differential Revision: [D22963408](https://our.internmc.facebook.com/intern/diff/D22963408)

[ghstack-poisoned]
